### DISCUSSION TITLE
Advice on getting valid transition targets for node

### DISCUSSION THREAD

**Discussion by mattpocock** [statelyai/xstate#1386] [General] [answered]
I'm working on [xstate-codegen](https://github.com/mattpocock/xstate-codegen), and I've been trying to think of a way to get all the valid transition targets from a node. I'd like to get them as strings, with the delimiter, so that I can use them for type checking.

Here's what I need to get:

1. The siblings of a node.
2. The children of all siblings, including their children.
3. Any nodes declared with an id property, to allow for '#id' transitions. I also need all of their children/grandchildren.

I am getting there, but I wondered if you have any shortcuts/utils I can use?

**Comment by davidkpiano** [comment, marked as answer]
Do you mean all the targets of all the transitions from a state node?

```js
function getAllTransitionTargets(stateNode) {
  return flatten(stateNode.transitions.map(t => {
    return t.target || [];
  }));
}
```

That should do it. You can dedupe using `Set()` and recursively get all the transition targets of the sibling nodes themselves (`Object.values(stateNode.states)`).

This is after the machine does the "processing" of the config object.

**Reply by mattpocock** [reply]
Sorry, I meant all the _possible_ transitions from a state node, not just the declared ones.

**Reply by Andarist** [reply]
Hm, could you clarify this a little bit more? Non-declared transitions are not valid and thus not possible. I know you know that so we must clearly just not understand each other correctly. Could you rephrase?

**Reply by mattpocock** [reply]
@Andarist Of course, my fault.

A loom for clarity:

https://www.loom.com/share/a3f9eb337338495e8f40c9553acce8c0

**Reply by Andarist** [reply]
There are no utils for this, as far as I know. Maybe we could include such in `@xstate/graph` package? although not 100% sure what's it exact purpose, it sounds like a good place for this and if not then a separate package with utils like this could be created

For now, I would just implement this in the codegen's codebase - we can later extract it. 

To get all ID targets you could probably do something like (just a pseudo code, not validated with real-life functions etc):
```js
import { getAllStateNodes } from 'xstate/lib/stateUtils'

const explicitIds = getAllStateNodes(machine).filter(node => node.config.id).map(node => node.id)
```

**Reply by davidkpiano** [reply]
Yeah, the possible targets would be a combination of the explicit IDs (as stated above) and the "relative" dot-representation of those IDs relative to that state node. All state nodes can share the `explicitIds`, so for the relative paths, you would filter the state nodes that are either _siblings_, _children_, or _cousins_ (children of siblings) and generate the relative paths.

**Reply by mattpocock** [reply]
Got this working, thanks for your help 👍
